app/http/httpclient.c

```diff
@@ -86,430 +81,460 @@ esp_isdigit( char c )
-		else
-			espconn_send( conn, (uint8_t *) req->post_data, strlen( req->post_data ) );
-		os_free( req->post_data );
-		req->post_data = NULL;
-	}
+  struct espconn  * conn  = (struct espconn *) arg;
+  request_args_t  * req = (request_args_t *) conn->reverse;
+
+  if ( req->post_data == NULL )
+  {
+    HTTPCLIENT_DEBUG( "All sent\n" );
+  }
+  else  
+  {
+    /* The headers were sent, now send the contents. */
+    HTTPCLIENT_DEBUG( "Sending request body\n" );
+    if ( req->secure )
+      espconn_secure_send( conn, (uint8_t *) req->post_data, strlen( req->post_data ) );
+    else
+      espconn_send( conn, (uint8_t *) req->post_data, strlen( req->post_data ) );
+    os_free( req->post_data );
+    req->post_data = NULL;
+  }
 }
 
 
 static void ICACHE_FLASH_ATTR http_connect_callback( void * arg )
 {
-	HTTPCLIENT_DEBUG( "Connected\n" );
-	struct espconn	* conn	= (struct espconn *) arg;
-	request_args_t	* req	= (request_args_t *) conn->reverse;
-
-	espconn_regist_recvcb( conn, http_receive_callback );
-	espconn_regist_sentcb( conn, http_send_callback );
-
-	char post_headers[32] = "";
-
-	if ( req->post_data != NULL ) /* If there is data then add Content-Length header. */
-	{
-		os_sprintf( post_headers, "Content-Length: %d\r\n", strlen( req->post_data ) );
-	}
-
-	if(req->headers == NULL) /* Avoid NULL pointer, it may cause exception */
-	{
-		req->headers = (char *)os_malloc(sizeof(char));
-		req->headers[0] = '\0';
-	}
-
-	char buf[69 + strlen( req->method ) + strlen( req->path ) + strlen( req->hostname ) +
-		 strlen( req->headers ) + strlen( post_headers )];
-	int len = os_sprintf( buf,
-			      "%s %s HTTP/1.1\r\n"
-			      "Host: %s:%d\r\n"
-			      "Connection: close\r\n"
-			      "User-Agent: ESP8266\r\n"
-			      "%s"
-			      "%s"
-			      "\r\n",
-			      req->method, req->path, req->hostname, req->port, req->headers, post_headers );
-
-	if ( req->secure )
-		espconn_secure_send( conn, (uint8_t *) buf, len );
-	else
-		espconn_send( conn, (uint8_t *) buf, len );
-	if(req->headers != NULL)
-		os_free( req->headers );
-	req->headers = NULL;
-	HTTPCLIENT_DEBUG( "Sending request header\n" );
+  HTTPCLIENT_DEBUG( "Connected\n" );
+  struct espconn  * conn  = (struct espconn *) arg;
+  request_args_t  * req = (request_args_t *) conn->reverse;
+
+  espconn_regist_recvcb( conn, http_receive_callback );
+  espconn_regist_sentcb( conn, http_send_callback );
+
+  char post_headers[32] = "";
+
+  if ( req->post_data != NULL ) /* If there is data then add Content-Length header. */
+  {
+    os_sprintf( post_headers, "Content-Length:%d\r\n", strlen( req->post_data ) );
+  }
+
+  if (req->headers == NULL) /* Avoid NULL pointer, it may cause exception */
+  {
+    req->headers = (char *)os_malloc(sizeof(char));
+    req->headers[0] = '\0';
+  }
+
+  char ua_header[32] = "";
+  int ua_len = 0;
+  
+  if ( os_strstr( req->headers, "User-Agent:" ) == NULL || os_strstr( req->headers, "user-agent:" ) == NULL)
```

Comment:
Shouldn't this be an AND instead of OR?

If `User-Agent:` AND `user-agent:` aren't present, then use default user agent.